### PR TITLE
Feature/course collections descriptions

### PR DESCRIPTION
**Pull request by eadm**
**YouTrack task**: [#APPS-1660](https://vyahhi.myjetbrains.com/youtrack/issue/APPS-1660), [#APPS-1683](https://vyahhi.myjetbrains.com/youtrack/issue/APPS-1683)

**Description List**:
* Add header to course collection list
* Fix crash on screen rotation in `CourseCollectionFragment`

### CODE REVIEW

File: app/src/main/java/org/stepic/droid/ui/adapters/CoursesAdapter.java [hunk 1]
```diff
@@ -61,9 +63,14 @@
     @NotNull
     private final DroppingPresenter droppingPresenter;
 
-    private int footerViewType = 1;
-    private int itemViewType = 2;
-    private final int NUMBER_OF_EXTRA_ITEMS;
+    private CoursesDescriptionContainer descriptionContainer;
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/ui/adapters/CoursesAdapter.java]: It is `nullable`

File: app/src/main/java/org/stepic/droid/ui/adapters/CoursesAdapter.java [hunk 2]
```diff
@@ -111,10 +118,18 @@ public CoursesAdapter(FragmentActivity activity,
...
     public CourseViewHolderBase onCreateViewHolder(ViewGroup parent, int viewType) {
-        if (viewType == footerViewType) {
+        if (viewType == HEADER_VIEW_TYPE) {
+            View view = inflater.inflate(R.layout.placeholder_view, parent, false);
+            ((RecyclerView.LayoutParams) view.getLayoutParams()).setMargins(
+                    -(int) contextActivity.getResources().getDimension(R.dimen.course_list_side_padding),
+                    -(int) contextActivity.getResources().getDimension(R.dimen.course_list_between_items_padding),
+                    -(int) contextActivity.getResources().getDimension(R.dimen.course_list_side_right_padding),
+                    0); // todo refactor layouts
+            return new HeaderItemViewHolder(view, descriptionContainer);
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/ui/adapters/CoursesAdapter.java]: `descriptionContainer` is nullable, but here it is `non nullable` implicitly.

File: app/src/main/java/org/stepic/droid/model/CoursesDescriptionContainer.kt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+package org.stepic.droid.model
+
+data class CoursesDescriptionContainer(
+        var description: String,
+        var colors: CollectionDescriptionColors
```
[the remark this answers is not in the document]
Comment (KirillMakarov) [reply app/src/main/java/org/stepic/droid/model/CoursesDescriptionContainer.kt]: I understood, why it was chosen as mutable. See comments below.

File: app/src/main/java/org/stepic/droid/ui/adapters/viewhoders/HeaderItemViewHolder.kt [hunk 4]
```diff
@@ -0,0 +1,16 @@
+package org.stepic.droid.ui.adapters.viewhoders
+
+import android.view.View
+import org.stepic.droid.model.CoursesDescriptionContainer
+import org.stepic.droid.ui.custom.PlaceholderTextView
+import org.stepic.droid.util.ColorUtil
+
+class HeaderItemViewHolder(view: View, private val descriptionContainer: CoursesDescriptionContainer): CourseViewHolderBase(view) {
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/ui/adapters/viewhoders/HeaderItemViewHolder.kt]: It is good chance to refactor logic of extending `CourseViewHolderBase`. Let's remove it and extend `RecyclerView.ViewHolder`. So, we can cast holders in `onBindView` and execute methods without strange mutable properties. It is relevant for `FooterItemViewHolder` too.

File: app/src/main/java/org/stepic/droid/ui/fragments/CourseCollectionFragment.kt [hunk 5]
```diff
@@ -12,11 +14,15 @@ class CourseCollectionFragment : CourseListFragmentBase() {
     companion object {
         private const val TITLE_KEY = "title_key"
         private const val COURSE_IDS = "course_ids"
+        private const val DESCRIPTION_TEXT = "description_text"
+        private const val DESCRIPTION_COLORS = "description_colors"
 
-        fun newInstance(title: String, courseIds: LongArray): CourseCollectionFragment {
+        fun newInstance(title: String, courseIds: LongArray, descriptionText: String, descriptionColors: CollectionDescriptionColors?): CourseCollectionFragment {
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/ui/fragments/CourseCollectionFragment.kt]: use `CoursesDescriptionContainer` here. We should make it parcelable

File: app/src/main/res/layout/placeholder_view.xml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<?xml version="1.0" encoding="utf-8"?>
+<org.stepic.droid.ui.custom.PlaceholderTextView
```
Comment (KirillMakarov) [inline app/src/main/res/layout/placeholder_view.xml]: Name of the file is not correct. It is related to collection header

File: app/src/main/java/org/stepic/droid/core/ScreenManagerImpl.java [hunk 7]
```diff
@@ -130,9 +131,10 @@ public void showNotifications(@NotNull Activity activity) {
     }
 
     @Override
-    public void showCoursesList(Activity activity, @NotNull CoursesCarouselInfo info) {
+    public void showCoursesList(Activity activity, @NotNull CoursesCarouselInfo info, @Nullable CollectionDescriptionColors descriptionColors) {
         Intent intent = new Intent(activity, CourseListActivity.class);
         intent.putExtra(CourseListActivity.COURSE_LIST_INFO_KEY, info);
+        intent.putExtra(CourseListActivity.COURSE_DESCRIPTION_COLORS, descriptionColors);
```
[the remark this answers is not in the document]
Comment (eadm) [reply app/src/main/java/org/stepic/droid/core/ScreenManagerImpl.java]: `CollectionDescriptionColors` is `enum`, so there is no `Parcelable` implementation.

File: app/src/main/java/org/stepic/droid/model/CoursesDescriptionContainer.kt [hunk 8]
```diff
@@ -1,6 +1,28 @@
...
 
+import android.os.Parcel
+import android.os.Parcelable
+
 data class CoursesDescriptionContainer(
         var description: String,
         var colors: CollectionDescriptionColors
-)
\ No newline at end of file
+) : Parcelable {
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/model/CoursesDescriptionContainer.kt]: Please, check Parcelable implementation like [this](https://github.com/StepicOrg/stepik-android/blob/6a3f3a91b2b9733737b31b534a1fe54e1f8fa13a/app/src/test/java/org/stepic/droid/model/code/ParcelableStringListTest.kt#L13-L13)

File: app/src/main/java/org/stepic/droid/ui/adapters/CoursesAdapter.java [hunk 9]
```diff
@@ -180,14 +195,14 @@ public void setDescriptionContainer(CoursesDescriptionContainer descriptionConta
             this.descriptionContainer = null;
             notifyItemRemoved(0);
         } else if (this.descriptionContainer != null) {
-            this.descriptionContainer.setDescription(descriptionContainer.getDescription());
-            this.descriptionContainer.setColors(descriptionContainer.getColors());
+//            this.descriptionContainer.setDescription(descriptionContainer.getDescription());
+//            this.descriptionContainer.setColors(descriptionContainer.getColors());
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/ui/adapters/CoursesAdapter.java]: why this code is commented?

File: app/src/main/java/org/stepic/droid/ui/adapters/viewhoders/FooterItemViewHolder.kt [hunk 10]
```diff
@@ -1,23 +1,16 @@
...
-            View.VISIBLE
-        } else {
-            View.GONE
-        }
-    }
-
-    companion object {
-        data class State(var isNeedShow: Boolean)
+    fun setLoaderVisibiluty(isNeedShow: Boolean) {
+        loadingRoot.changeVisibility(isNeedShow)
```
Comment (KirillMakarov) [inline app/src/main/java/org/stepic/droid/ui/adapters/viewhoders/FooterItemViewHolder.kt]: nice

**Pull request merged**